feat: Add did-webvh domain to protocol

3kz2501 opened the pull request:
Add new did method, did-webvh.

ibuibu commented on protocol/src/did_webvh/domain/crypto/hash.rs:
```diff
@@ -0,0 +1,56 @@
+    FailedToGenerateHash,
+}
+
+pub fn generate_multihash_with_base58encode(data: &[u8]) -> Result<String, CryptoError> {
```
base58_encodeの方が良い気がしました。ぱっと見で見つけたのでシングルコメントです🙏

tzskp1 reviewed (commented):
```diff
diff --git a/protocol/src/did_webvh/domain/crypto/hash.rs b/protocol/src/did_webvh/domain/crypto/hash.rs
index 7536b76..6cdd9f7 100644
--- a/protocol/src/did_webvh/domain/crypto/hash.rs
+++ b/protocol/src/did_webvh/domain/crypto/hash.rs
@@ -25,7 +25,7 @@ pub fn validate_hash(hash: &str) -> bool {
             Err(_) => false,
         },
         Err(e) => {
-            eprintln!("Error decoding hash: {:?}", e);
+            println!("Error decoding hash: {:?}", e);
             false
         }
     }
diff --git a/protocol/src/did_webvh/domain/did.rs b/protocol/src/did_webvh/domain/did.rs
index f0435ce..378e66d 100644
--- a/protocol/src/did_webvh/domain/did.rs
+++ b/protocol/src/did_webvh/domain/did.rs
@@ -1,85 +1,107 @@
 use super::crypto::hash::validate_hash;
 use const_format::formatcp;
-use serde::{Deserialize, Deserializer, Serialize, Serializer};
+use serde::{Deserialize, Deserializer, Serialize};
 use std::convert::TryFrom;
 use std::str::FromStr;
 
+// ref: https://www.w3.org/TR/did-core/#did-syntax
 static PAC_ENCODED: &str = r"%[0-9A-Fa-f]{2}";
 static IDCHAR: &str = formatcp!(r"[a-zA-Z0-9.\-_{PAC_ENCODED}]");
 static METHOD_CHAR: &str = r"[a-zA-Z0-9]";
-static DID_PAT: &str = formatcp!(
-    r"^did:(?<method_name>{METHOD_CHAR}+):(?<method_specific_id>({IDCHAR}*:)*{IDCHAR}+)$"
-);
+static METHOD_PAT: &str = formatcp!(r"{METHOD_CHAR}+");
+static METHOD_ID_PAT: &str = formatcp!(r"({IDCHAR}*:)*{IDCHAR}+");
+static DID_PAT: &str =
+    formatcp!(r"^did:(?<method_name>{METHOD_PAT}):(?<method_specific_id>{METHOD_ID_PAT})$");
 
-#[derive(Debug, Clone, PartialEq)]
-pub enum DidError {
-    InvalidMethod,
-    InvaliDidentifier,
-    InvalidFormat,
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize)]
+#[serde(transparent)]
+pub struct Did {
+    #[serde(skip_serializing)]
+    method_name_position: usize,
+    #[serde(skip_serializing)]
+    method_specific_id_position: usize,
+    inner: String,
 }
 
-#[derive(Debug, Clone, PartialEq, Eq)]
-pub struct Did {
-    method: String,
-    identifier: String,
+#[derive(Clone, Debug, PartialEq, thiserror::Error)]
+pub enum DidError {
+    #[error(transparent)]
+    Regex(#[from] regex::Error),
+    #[error("Failed to parse did")]
+    InvalidFormat,
+    #[error("Failed to parse method name")]
+    InvalidMethodName,
+    #[error("Failed to parse method specific id")]
+    InvalidMethodSpecificId,
 }
 
 impl Did {
     pub fn new(method: &str, identifier: &str) -> Result<Self, DidError> {
-        if method.is_empty() {
-            return Err(DidError::InvalidMethod);
+        let re_method = regex::Regex::new(METHOD_PAT)?;
+        let re_id = regex::Regex::new(METHOD_ID_PAT)?;
+        if !re_method.is_match(method) {
+            return Err(DidError::InvalidMethodName);
         }
-        if identifier.is_empty() {
-            return Err(DidError::InvaliDidentifier);
-        }
-        if !method.chars().all(|c| c.is_ascii_lowercase()) {
-            return Err(DidError::InvalidMethod);
+        if !re_id.is_match(identifier) {
+            return Err(DidError::InvalidMethodSpecificId);
         }
+        let method_name_position = 4 + method.len();
+        let method_specific_id_position = method_name_position + 1 + identifier.len();
         Ok(Self {
-            method: method.to_string(),
-            identifier: identifier.to_string(),
+            method_name_position,
+            method_specific_id_position,
+            inner: format!("did:{}:{}", method, identifier),
         })
     }
 
+    pub fn into_inner(self) -> String {
+        self.inner
+    }
+
     pub fn get_method(&self) -> &str {
-        &self.method
+        &self.inner[4..self.method_name_position]
     }
 
-    pub fn get_identifier(&self) -> &str {
-        &self.identifier
+    pub fn get_method_specific_id(&self) -> &str {
+        &self.inner[(self.method_name_position + 1)..self.method_specific_id_position]
     }
+}
 
-    pub fn to_string(&self) -> String {
-        format!("did:{}:{}", self.method, self.identifier)
+impl std::fmt::Display for Did {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.inner)
     }
 }
 
 impl From<DidWebvh> for Did {
     fn from(did: DidWebvh) -> Self {
-        Self {
-            method: did.get_did().get_method().to_string(),
-            identifier: did.get_did().get_identifier().to_string(),
-        }
+        did.did
     }
 }
 
+fn verify_did(s: &str) -> Result<(usize, usize), DidError> {
+    let re = regex::Regex::new(DID_PAT)?;
+    let caps = re.captures(s).ok_or(DidError::InvalidFormat)?;
+    Ok((
+        caps.name("method_name")
+            .ok_or(DidError::InvalidMethodName)?
+            .end(),
+        caps.name("method_specific_id")
+            .ok_or(DidError::InvalidMethodSpecificId)?
+            .end(),
+    ))
+}
+
 impl FromStr for Did {
     type Err = DidError;
 
     fn from_str(s: &str) -> Result<Self, Self::Err> {
-        let re = regex::Regex::new(DID_PAT).map_err(|_| DidError::InvalidFormat)?;
-        let caps = re.captures(s).ok_or(DidError::InvalidFormat)?;
-        let method = caps.name("method_name").unwrap().as_str();
-        let identifier = caps.name("method_specific_id").unwrap().as_str();
-        Self::new(method, identifier)
-    }
-}
-impl Serialize for Did {
-    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
-    where
-        S: Serializer,
-    {
-        serializer.serialize_str(&self.to_string())
+        let (method_name_position, method_specific_id_position) = verify_did(s)?;
+        Ok(Self {
+            method_name_position,
+            method_specific_id_position,
+            inner: s.into(),
+        })
     }
 }
 
@@ -89,8 +111,13 @@ impl<'de> Deserialize<'de> for Did {
         D: Deserializer<'de>,
     {
         let s = String::deserialize(deserializer)?;
-        s.parse()
-            .map_err(|_| serde::de::Error::custom("Invalid Did Format"))
+        let (method_name_position, method_specific_id_position) =
+            verify_did(&s).map_err(|_| serde::de::Error::custom("Invalid Did Format"))?;
+        Ok(Self {
+            method_name_position,
+            method_specific_id_position,
+            inner: s,
+        })
     }
 }
 
@@ -174,9 +201,9 @@ impl TryFrom<Did> for DidWebvh {
 
     fn try_from(did: Did) -> Result<Self, Self::Error> {
         if did.get_method() != "webvh" {
-            return Err(DidWebvhError::DidError(DidError::InvalidMethod));
+            return Err(DidWebvhError::DidError(DidError::InvalidMethodName));
         }
-        let parts: Vec<&str> = did.get_identifier().split(':').collect();
+        let parts: Vec<&str> = did.get_method_specific_id().split(':').collect();
         if parts.len() < 2 {
             return Err(DidWebvhError::InvalidPath);
         }
@@ -190,13 +217,13 @@ impl FromStr for DidWebvh {
     fn from_str(s: &str) -> Result<Self, Self::Err> {
         let did: Did = s.parse()?;
         if did.get_method() != "webvh" {
-            return Err(DidWebvhError::DidError(DidError::InvalidMethod));
+            return Err(DidWebvhError::DidError(DidError::InvalidMethodName));
         }
-        let parts: Vec<&str> = did.get_identifier().split(':').collect();
+        let parts: Vec<&str> = did.get_method_specific_id().split(':').collect();
         // parsed scid and domain
         // eg. ["scid", "example.com%3A8000", "path", "to", "resource"]
         if parts.len() < 2 {
-            return Err(DidWebvhError::DidError(DidError::InvaliDidentifier));
+            return Err(DidWebvhError::DidError(DidError::InvalidMethodSpecificId));
         }
         // domain may be have path
         Self::new(parts[0], parts[1..].join(":").as_str())
@@ -222,12 +249,12 @@ mod tests {
     fn test_did_example() -> Result<(), DidError> {
         let did = Did::new("example", "hogehoge")?;
         assert_eq!(did.get_method(), "example");
-        assert_eq!(did.get_identifier(), "hogehoge");
+        assert_eq!(did.get_method_specific_id(), "hogehoge");
         assert_eq!(did.to_string(), "did:example:hogehoge");
 
         let did = "did:example:hogehoge".parse::<Did>().unwrap();
         assert_eq!(did.get_method(), "example");
-        assert_eq!(did.get_identifier(), "hogehoge");
+        assert_eq!(did.get_method_specific_id(), "hogehoge");
         Ok(())
     }
 
@@ -239,7 +266,7 @@ mod tests {
         )?;
         assert_eq!(did.get_method(), "webvh");
         assert_eq!(
-            did.get_identifier(),
+            did.get_method_specific_id(),
             "QmdEjpG2gwEWZAx8YjBrw7mF1iuCqgrMh8S63M7PaC1Ldr:example.com:path:to:resource"
         );
         assert_eq!(
@@ -252,7 +279,7 @@ mod tests {
             .unwrap();
         assert_eq!(did.get_method(), "webvh");
         assert_eq!(
-            did.get_identifier(),
+            did.get_method_specific_id(),
             "QmdEjpG2gwEWZAx8YjBrw7mF1iuCqgrMh8S63M7PaC1Ldr:example.com%3A8000:path:to:resource"
         );
 
@@ -263,11 +290,11 @@ mod tests {
     fn test_did_error() {
         assert_eq!(
             Did::new("", "example").unwrap_err(),
-            DidError::InvalidMethod
+            DidError::InvalidMethodName
         );
         assert_eq!(
             Did::new("web", "").unwrap_err(),
-            DidError::InvaliDidentifier
+            DidError::InvalidMethodSpecificId
         );
     }
 
@@ -335,7 +362,7 @@ mod tests {
             "did:webvh:example.com%3A8000"
                 .parse::<DidWebvh>()
                 .unwrap_err(),
-            DidWebvhError::DidError(DidError::InvaliDidentifier)
+            DidWebvhError::DidError(DidError::InvalidMethodSpecificId)
         );
         assert_eq!(
             "did:webvh:QmdEjpG2gwEWZAx8YjBrw7mF1iuCqgrMh8S63M7PaC1Ldr:example.com%3A8000:path/to/resource"
diff --git a/protocol/src/did_webvh/domain/did_log_entry.rs b/protocol/src/did_webvh/domain/did_log_entry.rs
index 5d56cca..4f5f012 100644
--- a/protocol/src/did_webvh/domain/did_log_entry.rs
+++ b/protocol/src/did_webvh/domain/did_log_entry.rs
@@ -1,4 +1,4 @@
-use super::crypto::hash::generate_multihash_with_base58encode;
+use super::crypto::hash::generate_multihash_with_base58_encode;
 use super::did::{Did, DidWebvh};
 use super::did_document::DidDocument;
 use chrono::DateTime;
@@ -86,7 +86,7 @@ pub struct Parameters {
     pub ttl: Option<u32>,
 }
 
-fn verify_next_key_hashes(next_key_hashes: &Vec<String>) -> Result<(), ValidationError> {
+fn verify_next_key_hashes(next_key_hashes: &[String]) -> Result<(), ValidationError> {
     if next_key_hashes.is_empty() {
         return Err(ValidationError::new("Next key hashes must not be empty"));
     }
@@ -273,16 +273,16 @@ impl DidLogEntry {
         }
         let jcs = serde_json_canonicalizer::to_string(&entry)
             .map_err(|_| DidLogEntryError::InvalidFormat)?;
-        generate_multihash_with_base58encode(jcs.as_bytes())
+        generate_multihash_with_base58_encode(jcs.as_bytes())
             .map_err(|_| DidLogEntryError::FaildMultihash)
     }
 
     // calculate the next key hashes by the Update Keys from the previous entry.
-    pub fn calc_next_key_hash(&self, keys: &Vec<String>) -> Result<Self, DidLogEntryError> {
-        let generate_hashed_keys = |keys: &Vec<String>| {
+    pub fn calc_next_key_hash(&self, keys: &[String]) -> Result<Self, DidLogEntryError> {
+        let generate_hashed_keys = |keys: &[String]| {
             keys.iter()
                 .map(|key| {
-                    generate_multihash_with_base58encode(key.as_bytes())
+                    generate_multihash_with_base58_encode(key.as_bytes())
                         .map_err(|_| DidLogEntryError::FaildMultihash)
                 })
                 .collect::<Result<Vec<String>, DidLogEntryError>>()
@@ -320,7 +320,6 @@ mod tests {
             .map(|line| match serde_json::from_str(line) {
                 Ok(entry) => entry,
                 Err(e) => {
-                    eprintln!("Error: {:?}", e);
                     panic!("Error: {:?}", e);
                 }
             })
@@ -381,7 +380,7 @@ mod tests {
 
         entry.version_id = scid.to_string();
         entry.parameters.scid = Some(scid.to_string());
-        let identifier = entry.state.id.get_identifier();
+        let identifier = entry.state.id.get_method_specific_id();
         let identifier = identifier.replace("{SCID}", scid.as_str());
         entry.state.id = Did::new("webvh", identifier.as_str()).unwrap();
         println!("ENTRY_REPLACE_SCID: {:?}\n", entry);
```

I think this is a better way to do it.

3kz2501 commented:
> ```diff
> diff --git a/protocol/src/did_webvh/domain/crypto/hash.rs b/protocol/src/did_webvh/domain/crypto/hash.rs
> index 7536b76..6cdd9f7 100644
> --- a/protocol/src/did_webvh/domain/crypto/hash.rs
> +++ b/protocol/src/did_webvh/domain/crypto/hash.rs
> @@ -25,7 +25,7 @@ pub fn validate_hash(hash: &str) -> bool {
>              Err(_) => false,
>          },
>          Err(e) => {
> -            eprintln!("Error decoding hash: {:?}", e);
> +            println!("Error decoding hash: {:?}", e);
>              false
>          }
>      }
> diff --git a/protocol/src/did_webvh/domain/did.rs b/protocol/src/did_webvh/domain/did.rs
> index f0435ce..378e66d 100644
> --- a/protocol/src/did_webvh/domain/did.rs
> +++ b/protocol/src/did_webvh/domain/did.rs
> @@ -1,85 +1,107 @@
>  use super::crypto::hash::validate_hash;
>  use const_format::formatcp;
> -use serde::{Deserialize, Deserializer, Serialize, Serializer};
> +use serde::{Deserialize, Deserializer, Serialize};
>  use std::convert::TryFrom;
>  use std::str::FromStr;
>  
> +// ref: https://www.w3.org/TR/did-core/#did-syntax
>  static PAC_ENCODED: &str = r"%[0-9A-Fa-f]{2}";
>  static IDCHAR: &str = formatcp!(r"[a-zA-Z0-9.\-_{PAC_ENCODED}]");
>  static METHOD_CHAR: &str = r"[a-zA-Z0-9]";
> -static DID_PAT: &str = formatcp!(
> -    r"^did:(?<method_name>{METHOD_CHAR}+):(?<method_specific_id>({IDCHAR}*:)*{IDCHAR}+)$"
> -);
> +static METHOD_PAT: &str = formatcp!(r"{METHOD_CHAR}+");
> +static METHOD_ID_PAT: &str = formatcp!(r"({IDCHAR}*:)*{IDCHAR}+");
> +static DID_PAT: &str =
> +    formatcp!(r"^did:(?<method_name>{METHOD_PAT}):(?<method_specific_id>{METHOD_ID_PAT})$");
>  
> -#[derive(Debug, Clone, PartialEq)]
> -pub enum DidError {
> -    InvalidMethod,
> -    InvaliDidentifier,
> -    InvalidFormat,
> +#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize)]
> +#[serde(transparent)]
> +pub struct Did {
> +    #[serde(skip_serializing)]
> +    method_name_position: usize,
> +    #[serde(skip_serializing)]
> +    method_specific_id_position: usize,
> +    inner: String,
>  }
>  
> -#[derive(Debug, Clone, PartialEq, Eq)]
> -pub struct Did {
> -    method: String,
> -    identifier: String,
> +#[derive(Clone, Debug, PartialEq, thiserror::Error)]
> +pub enum DidError {
> +    #[error(transparent)]
> +    Regex(#[from] regex::Error),
> +    #[error("Failed to parse did")]
> +    InvalidFormat,
> +    #[error("Failed to parse method name")]
> +    InvalidMethodName,
> +    #[error("Failed to parse method specific id")]
> +    InvalidMethodSpecificId,
>  }
>  
>  impl Did {
>      pub fn new(method: &str, identifier: &str) -> Result<Self, DidError> {
> -        if method.is_empty() {
> -            return Err(DidError::InvalidMethod);
> +        let re_method = regex::Regex::new(METHOD_PAT)?;
> +        let re_id = regex::Regex::new(METHOD_ID_PAT)?;
> +        if !re_method.is_match(method) {
> +            return Err(DidError::InvalidMethodName);
>          }
> -        if identifier.is_empty() {
> -            return Err(DidError::InvaliDidentifier);
> -        }
> -        if !method.chars().all(|c| c.is_ascii_lowercase()) {
> -            return Err(DidError::InvalidMethod);
> +        if !re_id.is_match(identifier) {
> +            return Err(DidError::InvalidMethodSpecificId);
>          }
> +        let method_name_position = 4 + method.len();
> +        let method_specific_id_position = method_name_position + 1 + identifier.len();
>          Ok(Self {
> -            method: method.to_string(),
> -            identifier: identifier.to_string(),
> +            method_name_position,
> +            method_specific_id_position,
> +            inner: format!("did:{}:{}", method, identifier),
>          })
>      }
>  
> +    pub fn into_inner(self) -> String {
> +        self.inner
> +    }
> +
>      pub fn get_method(&self) -> &str {
> -        &self.method
> +        &self.inner[4..self.method_name_position]
>      }
>  
> -    pub fn get_identifier(&self) -> &str {
> -        &self.identifier
> +    pub fn get_method_specific_id(&self) -> &str {
> +        &self.inner[(self.method_name_position + 1)..self.method_specific_id_position]
>      }
> +}
>  
> -    pub fn to_string(&self) -> String {
> -        format!("did:{}:{}", self.method, self.identifier)
> +impl std::fmt::Display for Did {
> +    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
> +        write!(f, "{}", self.inner)
>      }
>  }
>  
>  impl From<DidWebvh> for Did {
>      fn from(did: DidWebvh) -> Self {
> -        Self {
> -            method: did.get_did().get_method().to_string(),
> -            identifier: did.get_did().get_identifier().to_string(),
> -        }
> +        did.did
>      }
>  }
>  
> +fn verify_did(s: &str) -> Result<(usize, usize), DidError> {
> +    let re = regex::Regex::new(DID_PAT)?;
> +    let caps = re.captures(s).ok_or(DidError::InvalidFormat)?;
> +    Ok((
> +        caps.name("method_name")
> +            .ok_or(DidError::InvalidMethodName)?
> +            .end(),
> +        caps.name("method_specific_id")
> +            .ok_or(DidError::InvalidMethodSpecificId)?
> +            .end(),
> +    ))
> +}
> +
>  impl FromStr for Did {
>      type Err = DidError;
>  
>      fn from_str(s: &str) -> Result<Self, Self::Err> {
> -        let re = regex::Regex::new(DID_PAT).map_err(|_| DidError::InvalidFormat)?;
> -        let caps = re.captures(s).ok_or(DidError::InvalidFormat)?;
> -        let method = caps.name("method_name").unwrap().as_str();
> -        let identifier = caps.name("method_specific_id").unwrap().as_str();
> -        Self::new(method, identifier)
> -    }
> -}
> -impl Serialize for Did {
> -    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
> -    where
> -        S: Serializer,
> -    {
> -        serializer.serialize_str(&self.to_string())
> +        let (method_name_position, method_specific_id_position) = verify_did(s)?;
> +        Ok(Self {
> +            method_name_position,
> +            method_specific_id_position,
> +            inner: s.into(),
> +        })
>      }
>  }
>  
> @@ -89,8 +111,13 @@ impl<'de> Deserialize<'de> for Did {
>          D: Deserializer<'de>,
>      {
>          let s = String::deserialize(deserializer)?;
> -        s.parse()
> -            .map_err(|_| serde::de::Error::custom("Invalid Did Format"))
> +        let (method_name_position, method_specific_id_position) =
> +            verify_did(&s).map_err(|_| serde::de::Error::custom("Invalid Did Format"))?;
> +        Ok(Self {
> +            method_name_position,
> +            method_specific_id_position,
> +            inner: s,
> +        })
>      }
>  }
>  
> @@ -174,9 +201,9 @@ impl TryFrom<Did> for DidWebvh {
>  
>      fn try_from(did: Did) -> Result<Self, Self::Error> {
>          if did.get_method() != "webvh" {
> -            return Err(DidWebvhError::DidError(DidError::InvalidMethod));
> +            return Err(DidWebvhError::DidError(DidError::InvalidMethodName));
>          }
> -        let parts: Vec<&str> = did.get_identifier().split(':').collect();
> +        let parts: Vec<&str> = did.get_method_specific_id().split(':').collect();
>          if parts.len() < 2 {
>              return Err(DidWebvhError::InvalidPath);
>          }
> @@ -190,13 +217,13 @@ impl FromStr for DidWebvh {
>      fn from_str(s: &str) -> Result<Self, Self::Err> {
>          let did: Did = s.parse()?;
>          if did.get_method() != "webvh" {
> -            return Err(DidWebvhError::DidError(DidError::InvalidMethod));
> +            return Err(DidWebvhError::DidError(DidError::InvalidMethodName));
>          }
> -        let parts: Vec<&str> = did.get_identifier().split(':').collect();
> +        let parts: Vec<&str> = did.get_method_specific_id().split(':').collect();
>          // parsed scid and domain
>          // eg. ["scid", "example.com%3A8000", "path", "to", "resource"]
>          if parts.len() < 2 {
> -            return Err(DidWebvhError::DidError(DidError::InvaliDidentifier));
> +            return Err(DidWebvhError::DidError(DidError::InvalidMethodSpecificId));
>          }
>          // domain may be have path
>          Self::new(parts[0], parts[1..].join(":").as_str())
> @@ -222,12 +249,12 @@ mod tests {
>      fn test_did_example() -> Result<(), DidError> {
>          let did = Did::new("example", "hogehoge")?;
>          assert_eq!(did.get_method(), "example");
> -        assert_eq!(did.get_identifier(), "hogehoge");
> +        assert_eq!(did.get_method_specific_id(), "hogehoge");
>          assert_eq!(did.to_string(), "did:example:hogehoge");
>  
>          let did = "did:example:hogehoge".parse::<Did>().unwrap();
>          assert_eq!(did.get_method(), "example");
> -        assert_eq!(did.get_identifier(), "hogehoge");
> +        assert_eq!(did.get_method_specific_id(), "hogehoge");
>          Ok(())
>      }
>  
> @@ -239,7 +266,7 @@ mod tests {
>          )?;
>          assert_eq!(did.get_method(), "webvh");
>          assert_eq!(
> -            did.get_identifier(),
> +            did.get_method_specific_id(),
>              "QmdEjpG2gwEWZAx8YjBrw7mF1iuCqgrMh8S63M7PaC1Ldr:example.com:path:to:resource"
>          );
>          assert_eq!(
> @@ -252,7 +279,7 @@ mod tests {
>              .unwrap();
>          assert_eq!(did.get_method(), "webvh");
>          assert_eq!(
> -            did.get_identifier(),
> +            did.get_method_specific_id(),
>              "QmdEjpG2gwEWZAx8YjBrw7mF1iuCqgrMh8S63M7PaC1Ldr:example.com%3A8000:path:to:resource"
>          );
>  
> @@ -263,11 +290,11 @@ mod tests {
>      fn test_did_error() {
>          assert_eq!(
>              Did::new("", "example").unwrap_err(),
> -            DidError::InvalidMethod
> +            DidError::InvalidMethodName
>          );
>          assert_eq!(
>              Did::new("web", "").unwrap_err(),
> -            DidError::InvaliDidentifier
> +            DidError::InvalidMethodSpecificId
>          );
>      }
>  
> @@ -335,7 +362,7 @@ mod tests {
>              "did:webvh:example.com%3A8000"
>                  .parse::<DidWebvh>()
>                  .unwrap_err(),
> -            DidWebvhError::DidError(DidError::InvaliDidentifier)
> +            DidWebvhError::DidError(DidError::InvalidMethodSpecificId)
>          );
>          assert_eq!(
>              "did:webvh:QmdEjpG2gwEWZAx8YjBrw7mF1iuCqgrMh8S63M7PaC1Ldr:example.com%3A8000:path/to/resource"
> diff --git a/protocol/src/did_webvh/domain/did_log_entry.rs b/protocol/src/did_webvh/domain/did_log_entry.rs
> index 5d56cca..4f5f012 100644
> --- a/protocol/src/did_webvh/domain/did_log_entry.rs
> +++ b/protocol/src/did_webvh/domain/did_log_entry.rs
> @@ -1,4 +1,4 @@
> -use super::crypto::hash::generate_multihash_with_base58encode;
> +use super::crypto::hash::generate_multihash_with_base58_encode;
>  use super::did::{Did, DidWebvh};
>  use super::did_document::DidDocument;
>  use chrono::DateTime;
> @@ -86,7 +86,7 @@ pub struct Parameters {
>      pub ttl: Option<u32>,
>  }
>  
> -fn verify_next_key_hashes(next_key_hashes: &Vec<String>) -> Result<(), ValidationError> {
> +fn verify_next_key_hashes(next_key_hashes: &[String]) -> Result<(), ValidationError> {
>      if next_key_hashes.is_empty() {
>          return Err(ValidationError::new("Next key hashes must not be empty"));
>      }
> @@ -273,16 +273,16 @@ impl DidLogEntry {
>          }
>          let jcs = serde_json_canonicalizer::to_string(&entry)
>              .map_err(|_| DidLogEntryError::InvalidFormat)?;
> -        generate_multihash_with_base58encode(jcs.as_bytes())
> +        generate_multihash_with_base58_encode(jcs.as_bytes())
>              .map_err(|_| DidLogEntryError::FaildMultihash)
>      }
>  
>      // calculate the next key hashes by the Update Keys from the previous entry.
> -    pub fn calc_next_key_hash(&self, keys: &Vec<String>) -> Result<Self, DidLogEntryError> {
> -        let generate_hashed_keys = |keys: &Vec<String>| {
> +    pub fn calc_next_key_hash(&self, keys: &[String]) -> Result<Self, DidLogEntryError> {
> +        let generate_hashed_keys = |keys: &[String]| {
>              keys.iter()
>                  .map(|key| {
> -                    generate_multihash_with_base58encode(key.as_bytes())
> +                    generate_multihash_with_base58_encode(key.as_bytes())
>                          .map_err(|_| DidLogEntryError::FaildMultihash)
>                  })
>                  .collect::<Result<Vec<String>, DidLogEntryError>>()
> @@ -320,7 +320,6 @@ mod tests {
>              .map(|line| match serde_json::from_str(line) {
>                  Ok(entry) => entry,
>                  Err(e) => {
> -                    eprintln!("Error: {:?}", e);
>                      panic!("Error: {:?}", e);
>                  }
>              })
> @@ -381,7 +380,7 @@ mod tests {
>  
>          entry.version_id = scid.to_string();
>          entry.parameters.scid = Some(scid.to_string());
> -        let identifier = entry.state.id.get_identifier();
> +        let identifier = entry.state.id.get_method_specific_id();
>          let identifier = identifier.replace("{SCID}", scid.as_str());
>          entry.state.id = Did::new("webvh", identifier.as_str()).unwrap();
>          println!("ENTRY_REPLACE_SCID: {:?}\n", entry);
> ```
> 
> I think this is a better way to do it.

Is it correct that the policy of this DID structure is that it is more cost-effective to have a single inner structure and acquire it by slice operation than to have multiple Methods and Identifiers as strings, in order to cope with memory limitations of edge devices?
My own opinion is that it would be fine for devices that are installed in the OS without worrying about such restrictions. I was thinking of having a String for each property for readability and so on. Also, Webvh also includes a Did, but I thought it would be troublesome to change the slice position for each placeholder conversion process performed at the time of SCID's Resolve process.
If the DID changes due to a change in the Method name or Domain, etc., there may be a need to recalculate the DID, and considering this, I am thinking that it may be better to have it as a String.

3kz2501 commented:
@tzskp1 

> ```diff
> diff --git a/protocol/src/did_webvh/domain/crypto/hash.rs b/protocol/src/did_webvh/domain/crypto/hash.rs
> index 7536b76..6cdd9f7 100644
> --- a/protocol/src/did_webvh/domain/crypto/hash.rs
> +++ b/protocol/src/did_webvh/domain/crypto/hash.rs
> @@ -25,7 +25,7 @@ pub fn validate_hash(hash: &str) -> bool {
>              Err(_) => false,
>          },
>          Err(e) => {
> -            eprintln!("Error decoding hash: {:?}", e);
> +            println!("Error decoding hash: {:?}", e);
>              false
>          }
>      }
> diff --git a/protocol/src/did_webvh/domain/did.rs b/protocol/src/did_webvh/domain/did.rs
> index f0435ce..378e66d 100644
> --- a/protocol/src/did_webvh/domain/did.rs
> +++ b/protocol/src/did_webvh/domain/did.rs
> @@ -1,85 +1,107 @@
>  use super::crypto::hash::validate_hash;
>  use const_format::formatcp;
> -use serde::{Deserialize, Deserializer, Serialize, Serializer};
> +use serde::{Deserialize, Deserializer, Serialize};
>  use std::convert::TryFrom;
>  use std::str::FromStr;
>  
> +// ref: https://www.w3.org/TR/did-core/#did-syntax
>  static PAC_ENCODED: &str = r"%[0-9A-Fa-f]{2}";
>  static IDCHAR: &str = formatcp!(r"[a-zA-Z0-9.\-_{PAC_ENCODED}]");
>  static METHOD_CHAR: &str = r"[a-zA-Z0-9]";
> -static DID_PAT: &str = formatcp!(
> -    r"^did:(?<method_name>{METHOD_CHAR}+):(?<method_specific_id>({IDCHAR}*:)*{IDCHAR}+)$"
> -);
> +static METHOD_PAT: &str = formatcp!(r"{METHOD_CHAR}+");
> +static METHOD_ID_PAT: &str = formatcp!(r"({IDCHAR}*:)*{IDCHAR}+");
> +static DID_PAT: &str =
> +    formatcp!(r"^did:(?<method_name>{METHOD_PAT}):(?<method_specific_id>{METHOD_ID_PAT})$");
>  
> -#[derive(Debug, Clone, PartialEq)]
> -pub enum DidError {
> -    InvalidMethod,
> -    InvaliDidentifier,
> -    InvalidFormat,
> +#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize)]
> +#[serde(transparent)]
> +pub struct Did {
> +    #[serde(skip_serializing)]
> +    method_name_position: usize,
> +    #[serde(skip_serializing)]
> +    method_specific_id_position: usize,
> +    inner: String,
>  }
>  
> -#[derive(Debug, Clone, PartialEq, Eq)]
> -pub struct Did {
> -    method: String,
> -    identifier: String,
> +#[derive(Clone, Debug, PartialEq, thiserror::Error)]
> +pub enum DidError {
> +    #[error(transparent)]
> +    Regex(#[from] regex::Error),
> +    #[error("Failed to parse did")]
> +    InvalidFormat,
> +    #[error("Failed to parse method name")]
> +    InvalidMethodName,
> +    #[error("Failed to parse method specific id")]
> +    InvalidMethodSpecificId,
>  }
>  
>  impl Did {
>      pub fn new(method: &str, identifier: &str) -> Result<Self, DidError> {
> -        if method.is_empty() {
> -            return Err(DidError::InvalidMethod);
> +        let re_method = regex::Regex::new(METHOD_PAT)?;
> +        let re_id = regex::Regex::new(METHOD_ID_PAT)?;
> +        if !re_method.is_match(method) {
> +            return Err(DidError::InvalidMethodName);
>          }
> -        if identifier.is_empty() {
> -            return Err(DidError::InvaliDidentifier);
> -        }
> -        if !method.chars().all(|c| c.is_ascii_lowercase()) {
> -            return Err(DidError::InvalidMethod);
> +        if !re_id.is_match(identifier) {
> +            return Err(DidError::InvalidMethodSpecificId);
>          }
> +        let method_name_position = 4 + method.len();
> +        let method_specific_id_position = method_name_position + 1 + identifier.len();
>          Ok(Self {
> -            method: method.to_string(),
> -            identifier: identifier.to_string(),
> +            method_name_position,
> +            method_specific_id_position,
> +            inner: format!("did:{}:{}", method, identifier),
>          })
>      }
>  
> +    pub fn into_inner(self) -> String {
> +        self.inner
> +    }
> +
>      pub fn get_method(&self) -> &str {
> -        &self.method
> +        &self.inner[4..self.method_name_position]
>      }
>  
> -    pub fn get_identifier(&self) -> &str {
> -        &self.identifier
> +    pub fn get_method_specific_id(&self) -> &str {
> +        &self.inner[(self.method_name_position + 1)..self.method_specific_id_position]
>      }
> +}
>  
> -    pub fn to_string(&self) -> String {
> -        format!("did:{}:{}", self.method, self.identifier)
> +impl std::fmt::Display for Did {
> +    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
> +        write!(f, "{}", self.inner)
>      }
>  }
>  
>  impl From<DidWebvh> for Did {
>      fn from(did: DidWebvh) -> Self {
> -        Self {
> -            method: did.get_did().get_method().to_string(),
> -            identifier: did.get_did().get_identifier().to_string(),
> -        }
> +        did.did
>      }
>  }
>  
> +fn verify_did(s: &str) -> Result<(usize, usize), DidError> {
> +    let re = regex::Regex::new(DID_PAT)?;
> +    let caps = re.captures(s).ok_or(DidError::InvalidFormat)?;
> +    Ok((
> +        caps.name("method_name")
> +            .ok_or(DidError::InvalidMethodName)?
> +            .end(),
> +        caps.name("method_specific_id")
> +            .ok_or(DidError::InvalidMethodSpecificId)?
> +            .end(),
> +    ))
> +}
> +
>  impl FromStr for Did {
>      type Err = DidError;
>  
>      fn from_str(s: &str) -> Result<Self, Self::Err> {
> -        let re = regex::Regex::new(DID_PAT).map_err(|_| DidError::InvalidFormat)?;
> -        let caps = re.captures(s).ok_or(DidError::InvalidFormat)?;
> -        let method = caps.name("method_name").unwrap().as_str();
> -        let identifier = caps.name("method_specific_id").unwrap().as_str();
> -        Self::new(method, identifier)
> -    }
> -}
> -impl Serialize for Did {
> -    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
> -    where
> -        S: Serializer,
> -    {
> -        serializer.serialize_str(&self.to_string())
> +        let (method_name_position, method_specific_id_position) = verify_did(s)?;
> +        Ok(Self {
> +            method_name_position,
> +            method_specific_id_position,
> +            inner: s.into(),
> +        })
>      }
>  }
>  
> @@ -89,8 +111,13 @@ impl<'de> Deserialize<'de> for Did {
>          D: Deserializer<'de>,
>      {
>          let s = String::deserialize(deserializer)?;
> -        s.parse()
> -            .map_err(|_| serde::de::Error::custom("Invalid Did Format"))
> +        let (method_name_position, method_specific_id_position) =
> +            verify_did(&s).map_err(|_| serde::de::Error::custom("Invalid Did Format"))?;
> +        Ok(Self {
> +            method_name_position,
> +            method_specific_id_position,
> +            inner: s,
> +        })
>      }
>  }
>  
> @@ -174,9 +201,9 @@ impl TryFrom<Did> for DidWebvh {
>  
>      fn try_from(did: Did) -> Result<Self, Self::Error> {
>          if did.get_method() != "webvh" {
> -            return Err(DidWebvhError::DidError(DidError::InvalidMethod));
> +            return Err(DidWebvhError::DidError(DidError::InvalidMethodName));
>          }
> -        let parts: Vec<&str> = did.get_identifier().split(':').collect();
> +        let parts: Vec<&str> = did.get_method_specific_id().split(':').collect();
>          if parts.len() < 2 {
>              return Err(DidWebvhError::InvalidPath);
>          }
> @@ -190,13 +217,13 @@ impl FromStr for DidWebvh {
>      fn from_str(s: &str) -> Result<Self, Self::Err> {
>          let did: Did = s.parse()?;
>          if did.get_method() != "webvh" {
> -            return Err(DidWebvhError::DidError(DidError::InvalidMethod));
> +            return Err(DidWebvhError::DidError(DidError::InvalidMethodName));
>          }
> -        let parts: Vec<&str> = did.get_identifier().split(':').collect();
> +        let parts: Vec<&str> = did.get_method_specific_id().split(':').collect();
>          // parsed scid and domain
>          // eg. ["scid", "example.com%3A8000", "path", "to", "resource"]
>          if parts.len() < 2 {
> -            return Err(DidWebvhError::DidError(DidError::InvaliDidentifier));
> +            return Err(DidWebvhError::DidError(DidError::InvalidMethodSpecificId));
>          }
>          // domain may be have path
>          Self::new(parts[0], parts[1..].join(":").as_str())
> @@ -222,12 +249,12 @@ mod tests {
>      fn test_did_example() -> Result<(), DidError> {
>          let did = Did::new("example", "hogehoge")?;
>          assert_eq!(did.get_method(), "example");
> -        assert_eq!(did.get_identifier(), "hogehoge");
> +        assert_eq!(did.get_method_specific_id(), "hogehoge");
>          assert_eq!(did.to_string(), "did:example:hogehoge");
>  
>          let did = "did:example:hogehoge".parse::<Did>().unwrap();
>          assert_eq!(did.get_method(), "example");
> -        assert_eq!(did.get_identifier(), "hogehoge");
> +        assert_eq!(did.get_method_specific_id(), "hogehoge");
>          Ok(())
>      }
>  
> @@ -239,7 +266,7 @@ mod tests {
>          )?;
>          assert_eq!(did.get_method(), "webvh");
>          assert_eq!(
> -            did.get_identifier(),
> +            did.get_method_specific_id(),
>              "QmdEjpG2gwEWZAx8YjBrw7mF1iuCqgrMh8S63M7PaC1Ldr:example.com:path:to:resource"
>          );
>          assert_eq!(
> @@ -252,7 +279,7 @@ mod tests {
>              .unwrap();
>          assert_eq!(did.get_method(), "webvh");
>          assert_eq!(
> -            did.get_identifier(),
> +            did.get_method_specific_id(),
>              "QmdEjpG2gwEWZAx8YjBrw7mF1iuCqgrMh8S63M7PaC1Ldr:example.com%3A8000:path:to:resource"
>          );
>  
> @@ -263,11 +290,11 @@ mod tests {
>      fn test_did_error() {
>          assert_eq!(
>              Did::new("", "example").unwrap_err(),
> -            DidError::InvalidMethod
> +            DidError::InvalidMethodName
>          );
>          assert_eq!(
>              Did::new("web", "").unwrap_err(),
> -            DidError::InvaliDidentifier
> +            DidError::InvalidMethodSpecificId
>          );
>      }
>  
> @@ -335,7 +362,7 @@ mod tests {
>              "did:webvh:example.com%3A8000"
>                  .parse::<DidWebvh>()
>                  .unwrap_err(),
> -            DidWebvhError::DidError(DidError::InvaliDidentifier)
> +            DidWebvhError::DidError(DidError::InvalidMethodSpecificId)
>          );
>          assert_eq!(
>              "did:webvh:QmdEjpG2gwEWZAx8YjBrw7mF1iuCqgrMh8S63M7PaC1Ldr:example.com%3A8000:path/to/resource"
> diff --git a/protocol/src/did_webvh/domain/did_log_entry.rs b/protocol/src/did_webvh/domain/did_log_entry.rs
> index 5d56cca..4f5f012 100644
> --- a/protocol/src/did_webvh/domain/did_log_entry.rs
> +++ b/protocol/src/did_webvh/domain/did_log_entry.rs
> @@ -1,4 +1,4 @@
> -use super::crypto::hash::generate_multihash_with_base58encode;
> +use super::crypto::hash::generate_multihash_with_base58_encode;
>  use super::did::{Did, DidWebvh};
>  use super::did_document::DidDocument;
>  use chrono::DateTime;
> @@ -86,7 +86,7 @@ pub struct Parameters {
>      pub ttl: Option<u32>,
>  }
>  
> -fn verify_next_key_hashes(next_key_hashes: &Vec<String>) -> Result<(), ValidationError> {
> +fn verify_next_key_hashes(next_key_hashes: &[String]) -> Result<(), ValidationError> {
>      if next_key_hashes.is_empty() {
>          return Err(ValidationError::new("Next key hashes must not be empty"));
>      }
> @@ -273,16 +273,16 @@ impl DidLogEntry {
>          }
>          let jcs = serde_json_canonicalizer::to_string(&entry)
>              .map_err(|_| DidLogEntryError::InvalidFormat)?;
> -        generate_multihash_with_base58encode(jcs.as_bytes())
> +        generate_multihash_with_base58_encode(jcs.as_bytes())
>              .map_err(|_| DidLogEntryError::FaildMultihash)
>      }
>  
>      // calculate the next key hashes by the Update Keys from the previous entry.
> -    pub fn calc_next_key_hash(&self, keys: &Vec<String>) -> Result<Self, DidLogEntryError> {
> -        let generate_hashed_keys = |keys: &Vec<String>| {
> +    pub fn calc_next_key_hash(&self, keys: &[String]) -> Result<Self, DidLogEntryError> {
> +        let generate_hashed_keys = |keys: &[String]| {
>              keys.iter()
>                  .map(|key| {
> -                    generate_multihash_with_base58encode(key.as_bytes())
> +                    generate_multihash_with_base58_encode(key.as_bytes())
>                          .map_err(|_| DidLogEntryError::FaildMultihash)
>                  })
>                  .collect::<Result<Vec<String>, DidLogEntryError>>()
> @@ -320,7 +320,6 @@ mod tests {
>              .map(|line| match serde_json::from_str(line) {
>                  Ok(entry) => entry,
>                  Err(e) => {
> -                    eprintln!("Error: {:?}", e);
>                      panic!("Error: {:?}", e);
>                  }
>              })
> @@ -381,7 +380,7 @@ mod tests {
>  
>          entry.version_id = scid.to_string();
>          entry.parameters.scid = Some(scid.to_string());
> -        let identifier = entry.state.id.get_identifier();
> +        let identifier = entry.state.id.get_method_specific_id();
>          let identifier = identifier.replace("{SCID}", scid.as_str());
>          entry.state.id = Did::new("webvh", identifier.as_str()).unwrap();
>          println!("ENTRY_REPLACE_SCID: {:?}\n", entry);
> ```
> 
> I think this is a better way to do it.

Is it correct that the policy of this DID structure is that it is more cost-effective to have a single inner structure and acquire it by slice operation than to have multiple Methods and Identifiers as strings, in order to cope with memory limitations of edge devices?
My own opinion is that it would be fine for devices that are installed in the OS without worrying about such restrictions. I was thinking of having a String for each property for readability and so on. Also, Webvh also includes a Did, but I thought it would be troublesome to change the slice position for each placeholder conversion process performed at the time of SCID's Resolve process.
If the DID changes due to a change in the Method name or Domain, etc., there may be a need to recalculate the DID, and considering this, I am thinking that it may be better to have it as a String.

tzskp1 commented:
@3kz2501 
> Is it correct that the policy of this DID structure is that it is more cost-effective to have a single inner structure and acquire it by slice operation than to have multiple Methods and Identifiers as strings, in order to cope with memory limitations of edge devices?

Yes. In addition I think my code fits the way of Rust.

I don't think the implementation details of the DID entities need to be readable.
The ideal encapsulation would be to look only at the interface and not worry about the implementation details when changing the DIDWebVh implementation. 
And my guess is that the position will not shift unless the DID specifications are changed.

What exactly is the point where it is difficult to read?

tzskp1 commented:
@3kz2501 
Oh, I see.
But even in that case, we should create a setter, and we have to do validation in the setter.
I think it is possible to create a temporary type as you mentioned to avoid validation cost.

3kz2501 commented on protocol/src/did_webvh/domain/did.rs:
```diff
@@ -0,0 +1,385 @@
+    }
+}
+
+impl Serialize for Did {
```
@tzskp1 I had a suggestion to remove this one and add `#[serde(skip_serializing)]`, but when I tested it, it did not work and was not skipped, so I have not applied that suggestion only.

3kz2501 commented:
@tzskp1 @ibuibu すみません, Conflict やLint Error など解消してApprove 外れたため再度送らせていただきました
